crates/huub/src/constraints/int_diffn.rs

```diff
@@ -0,0 +1,829 @@
+			.collect();
+
+		let box_size: Vec<Vec<_>> = self
+			.box_size
+			.iter()
+			.map(|row| row.iter().map(|v| slv.get_solver_int(*v)).collect())
+			.collect();
+		IntDiffnSweep::new_in(slv, box_pos, box_size, self.non_strict);
+		Ok(())
+	}
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+/// Active forbidden regions
+struct ForbiddenRegion {
+	/// lower bound of each dimension
+	lb: SmallVec<[IntVal; 3]>,
+	/// upper bound of each dimension
+	ub: SmallVec<[IntVal; 3]>,
+}
+
+impl IntDiffnSweep {
+	/// Prepare a new [`IntDiffnSweep`] propagator to be posted to the
+	/// solver.
+	pub fn new_in<P: PropagatorInitActions + ?Sized>(
+		solver: &mut P,
+		box_posn: Vec<Vec<IntView>>,
+		box_size: Vec<Vec<IntView>>,
+		non_strict: bool,
+	) {
+		// Make sure all sizes are fixed before enqueueing
+		let fixed = box_size
+			.iter()
+			.flatten()
+			.all(|&v| solver.get_int_val(v).is_some());
+
+		let mut box_posn_prop: Vec<Vec<IntView>> = box_posn.clone();
+		let box_size_prop: Vec<Vec<IntView>> = box_size.clone();
+
+		// Do not consider objects that are fixed and have a size of 0 when non_strict
+		if non_strict && fixed {
+			let mut box_size_fixed: Vec<_> = box_size
+				.iter()
+				.map(|row| {
+					row.iter().map(|&v| {
+						if solver.get_int_val(v) == Some(0) {
+							1
+						} else {
+							0
+						}
+					})
+				})
+				.collect();
+
+			let contains_zero: Vec<usize> = box_size_fixed
+				.iter_mut()
+				.map(|row| if row.contains(&1) { 1 } else { 0 })
+				.collect();
+
+			box_posn_prop = box_posn_prop
+				.into_iter()
+				.enumerate()
+				.filter(|(i, _)| contains_zero[*i] == 0)
+				.map(|(_, row)| row)
+				.collect();
+
+			box_posn_prop = box_posn_prop
+				.into_iter()
+				.enumerate()
+				.filter(|(i, _)| contains_zero[*i] == 0)
+				.map(|(_, row)| row)
+				.collect();
+		}
+
+		// Tracks whether an rectangles posn domains are fixed
+		let fixed_trail = (0..box_size.len())
+			.map(|_| solver.new_trailed_int(0))
+			.collect();
+		let remove_trail = (0..box_size.len())
+			.map(|_| solver.new_trailed_int(0))
+			.collect();
+		let lb_sizes_prop = vec![vec![0; box_posn[0].len()]; box_posn.len()];
+		let ub_tracker_prop = vec![vec![0; box_posn[0].len()]; box_posn.len()];
+		let lb_tracker_prop = vec![vec![0; box_posn[0].len()]; box_posn.len()];
+
+		let prop = solver.add_propagator(
+			Box::new(Self {
+				box_posn: box_posn_prop,
+				box_size: box_size_prop,
+				dimensions: box_posn[0].len(),
+				target: fixed_trail,
+				source: remove_trail,
+				fixed_sizes: fixed,
+				ub_tracker: ub_tracker_prop,
+				lb_tracker: lb_tracker_prop,
+				lb_sizes: lb_sizes_prop,
+				non_strict,
+			}),
+			PriorityLevel::Lowest,
+		);
+
+		for v in box_posn.into_iter().flatten() {
+			solver.enqueue_on_int_change(prop, v, IntPropCond::Bounds);
+		}
+        solver.enqueue_now(prop);
+	}
+
+	/// Prune the lower bounds of the domain
+	fn prune_min<P: PropagationActions>(
```

Comment:
It would be great if for the (inner) propagation methods (like `prune_min`, `prune_max`, `adjust_sweep_min`, `generate_fr`, etc.) we could add comments that really explain the reasoning, not just the actions being taken. Can you adjust them so that the method says what and how it's going to prune? Then the comments in the method should talk about why things are done and what assumptions are made at different parts of the code.